### PR TITLE
Polish code in preparation for a Groovy upgrade

### PR DESCRIPTION
**Pull request by paplorinc**
A step towards #8681

This PR doesn't change the Groovy version anymore, it just fixes a few quirks that would fail with 2.5.7

See:
* http://groovy-lang.org/changelogs/changelog-2.5.5.html
* http://groovy-lang.org/changelogs/changelog-2.5.6.html
* http://groovy-lang.org/changelogs/changelog-2.5.7.html

Groovy bugs reported:

- [ ] https://issues.apache.org/jira/browse/GROOVY-9167
- [ ] https://issues.apache.org/jira/browse/GROOVY-9166
- [x] https://issues.apache.org/jira/browse/GROOVY-9171

### CODE REVIEW

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy [hunk 1]
```diff
@@ -48,7 +48,7 @@ class MainTest extends Specification {
             [complete: { completedSuccessfully = true }, completeWithFailure: { completedWithFailure = true; failure = it }] as ExecutionCompleter
         }
 
-        protected CommandLineActionFactory createActionFactory() {
+        CommandLineActionFactory createActionFactory() {
```
Comment (lptr) [inline subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy]: Why this change?

File: subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts [hunk 2]
```diff
@@ -26,7 +26,7 @@ dependencies {
 
 // tag::pom-dependencies[]
 dependencies {
-    testImplementation("org.codehaus.groovy:groovy-all:2.5.4")
+    testImplementation("org.codehaus.groovy:groovy-all:2.5.7")
```
Comment (lptr) [inline subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts]: I'd say we shouldn't upgrade samples versions randomly. Instead, we should tackle this: https://github.com/gradle/gradle/issues/6044

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy (hunk 1 above)
Comment (paplorinc) [reply subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy]: it overrides a public method

File: subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts (hunk 2 above)
Comment (paplorinc) [reply subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts]: Upgraded, to make sure they pass and that I don't leave them behind, i.e. create more chores.
Would you like me to revert these?

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy]: Is this something you discovered accidentally, or something that Groovy 2.5.7 now fails because of?

File: subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts (hunk 2 above)
Comment (lptr) [reply subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts]: I think it's okay either way. But we should push forward with doing this automatically.

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy (hunk 1 above)
Comment (paplorinc) [reply subprojects/launcher/src/test/groovy/org/gradle/launcher/MainTest.groovy]: It's a failure now

File: subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts (hunk 2 above)
Comment (paplorinc) [reply subprojects/docs/src/samples/userguide/mavenMigration/basic/kotlin/build.gradle.kts]: Absolutely!

File: subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/FingerprinterFixture.groovy [hunk 3]
```diff
@@ -31,28 +31,46 @@ import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
...
-    private fileHasher = new TestFileHasher()
-    private stringInterner = new StringInterner()
-    private fileSystem = TestFiles.fileSystem()
-    private fileSystemMirror = new DefaultFileSystemMirror(new NoWellKnownFileLocations())
+    private inputFingerprinter
+    private outputFingerprinter
 
-    final inputFingerprinter = new AbsolutePathFileCollectionFingerprinter(
-        new DefaultFileSystemSnapshotter(
-            fileHasher,
```
Comment (paplorinc) [inline subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/FingerprinterFixture.groovy]: trait references to other fields aren't atomic and see their values as `null`

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 4]
```diff
@@ -210,21 +210,6 @@ include 'a', 'b'
         succeeds ":b:${language.compileTaskName}"
         executedAndNotSkipped ":a:${language.compileTaskName}", ":b:${language.compileTaskName}"
 
-        when:
-        // add public constructor to replace the default, should not change
```
Comment (paplorinc) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: * implicit and explicit default constructors don't generate the same bytecode anymore (implicit one is marked as `@Generated` now), see: https://issues.apache.org/jira/browse/GROOVY-9051
* static initializers generate and empty if statement. This seems like a bug, opened:
https://issues.apache.org/jira/browse/GROOVY-9166

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 5]
```diff
@@ -314,22 +300,7 @@ public class ToolImpl {
...
-            }
-"""
-
-        then:
-        succeeds ":b:${language.compileTaskName}"
-        executedAndNotSkipped ":a:${language.compileTaskName}"
-        skipped ":b:${language.compileTaskName}"
-
-        when:
-        // add static initializer and constructor
```
Comment (paplorinc) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: @wolfs, do we actually care about the empty constructor usecase, or was it just an easily testable dummy change?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TransformationNodeSpec.groovy [hunk 6]
```diff
@@ -72,13 +72,13 @@ class TransformationNodeSpec extends Specification {
...
 
         then:
-        1 * hardSuccessor.execute(initialNode)
-        1 * graphDependenciesResolver.computeDependencyNodes(transformationStep) >> container
-        1 * dependencyResolver.resolveDependenciesFor(null, container) >> [additionalNode]
-        1 * hardSuccessor.execute(additionalNode)
         1 * transformationStep.dependencies >> transformerDependencies
         1 * dependencyResolver.resolveDependenciesFor(null, transformerDependencies) >> [isolationNode]
         1 * hardSuccessor.execute(isolationNode)
+        1 * graphDependenciesResolver.computeDependencyNodes(transformationStep) >> container
```
Comment (paplorinc) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TransformationNodeSpec.groovy]: order matters now (Thanks @bigdaz)

Comment (wolfs) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TransformationNodeSpec.groovy]: Is this documented somewhere?

Comment (paplorinc) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TransformationNodeSpec.groovy]: Unfortunately I couldn't find it, I'm not even sure what Groovy has to do here, I thought this was Spock magic...

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 4 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: TODO separate Groovy tests

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 5 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Separated to Groovy & Java test suite (changed in different commits for reviewability)

File: buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy [hunk 7]
```diff
@@ -30,8 +30,8 @@ class RebaselinePerformanceTests extends SourceTask {
 
     @TaskAction
     void rebaseline() {
-        getSource().each { file ->
-            file.text = rebaseline(file.text, baseline)
+        for (file in getSource()) {
```
Comment (paplorinc) [inline buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy]: Filed: https://issues.apache.org/jira/browse/GROOVY-9171

File: buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy [hunk 8]
```diff
@@ -40,7 +40,7 @@ class RebaselinePerformanceTests extends SourceTask {
         this.baseline = baseline
     }
 
-    static String rebaseline(String fileContent, String baseline) {
+    private static String rebaselineContent(String fileContent, String baseline) {
```
Comment (wolfs) [inline buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy]: This should not be private when it is referenced from the test.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 9]
```diff
@@ -215,4 +215,174 @@ public @interface MyAnnotation { }
...
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        skipped ":b:${language.compileTaskName}"
+
+        when:
+        // add public method, should change
+        sourceFile.text = """
+            public class ToolImpl { 
+                public void execute() { String s = toString(); } 
+            }
+"""
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Nit: Can you let the closing `"""` have the same indentation as the opening ones?

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 10]
```diff
@@ -215,4 +215,174 @@ public @interface MyAnnotation { }
         executedAndNotSkipped ":b:compileGroovy"
         failure.assertHasCause('Bad AST transformation!')
     }
+
+    def "doesn't recompile when private element of implementation class changes"() {
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Can we maybe leave most of the test shared between Groovy and Java? I.e. the parts where we do the common changes? And then only add an extra test for the constructor changes to Java?

File: buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy (hunk 8 above)
Comment (lptr) [reply buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy]: If you want to signal that it's not `private` because of testing then `@VisibleForTesting` might be a good option.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 10 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Done - also added normal initializer changes to the Java compile avoidance test

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 9 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Done

File: buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy (hunk 8 above)
Comment (paplorinc) [reply buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/RebaselinePerformanceTests.groovy]: Done

**Comment by paplorinc** [comment]
Force pushed to remove invalid commits

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 11]
```diff
@@ -24,49 +24,49 @@ abstract class AbstractGroovyCompileAvoidanceIntegrationSpec extends AbstractJav
...
-@GroovyASTTransformation(phase = CompilePhase.CANONICALIZATION)
-public class MyASTTransformation extends AbstractASTTransformation {
-    @Override
-    public void visit(ASTNode[] nodes, SourceUnit source) {
-        assert false: "Bad AST transformation!"
-    }
-    public void foo() { }
-}
-    """
+        import org.codehaus.groovy.transform.*;
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: In order to be consistent with the other ones, this needs to be indented once more, right?

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 12]
```diff
@@ -86,11 +86,11 @@ public class MyASTTransformation extends AbstractASTTransformation {
 
     private String astTransformationAnnotation() {
         """
-import java.lang.annotation.*;
-@Retention(RetentionPolicy.RUNTIME)
-@Target([ElementType.TYPE])
-@org.codehaus.groovy.transform.GroovyASTTransformationClass("MyASTTransformation")
-public @interface MyAnnotation { }
+        import java.lang.annotation.*;
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: This also should be indented once more.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 13]
```diff
@@ -615,16 +566,16 @@ public class ToolImpl {
...
-    void a() { 
-        b(); 
-        String c = c(); 
-    } 
-    @Override String c() { 
-        return null; 
-    } 
-}
-"""
+        public class A extends B {
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Indent one more.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 14]
```diff
@@ -673,15 +624,15 @@ public class A extends B {
...
-public class A extends B { 
-    void a() { 
-        b(); 
-    } 
-    @Override String d() { 
-        return null; 
-    } 
-}
-"""
+        public class A extends B {
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Indent one more.

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 15]
```diff
@@ -293,39 +276,7 @@ public class ToolImpl {
...
-            }
-"""
-
-        then:
-        succeeds ":b:${language.compileTaskName}"
-        executedAndNotSkipped ":a:${language.compileTaskName}"
-        skipped ":b:${language.compileTaskName}"
-
-        when:
-        // change static initializer and constructor
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Does changing an already existing static initializer and constructor affect Groovy compilation? I would expect that the task stays up-to-date as well. We could add this test to the base class as a separate test. Or do we have that coverage already?

Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Ok, sure, separated

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 14 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Done

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 12 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Done

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 11 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy]: Indeed, thanks

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy (hunk 13 above)
Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: Done

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 16]
```diff
@@ -293,39 +276,55 @@ public class ToolImpl {
...
         then:
         succeeds ":b:${language.compileTaskName}"
         executedAndNotSkipped ":a:${language.compileTaskName}"
         skipped ":b:${language.compileTaskName}"
+    }
 
-        when:
-        // add static initializer and constructor
-        sourceFile.text = """
+    def "doesn't recompile when empty initializer, static initializer or constructor is added"() {
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]:
```suggestion
    def "doesn't recompile when initializer, static initializer or constructor is changed"() {
```

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy [hunk 17]
```diff
@@ -293,39 +276,55 @@ public class ToolImpl {
...
+        sourceFile << """
             public class ToolImpl {
-                static { }
-                public ToolImpl() { }
-                public Object s = "12";
-                public void execute() { String s = toString(); }
+                {}
+                static {}
+                public ToolImpl() {}
+                public Object s = String.valueOf(12);
```
Comment (wolfs) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: `s` and the `execute` method are not interesting for this test any more and can be removed. If they are not removed, they probably shouldn't be changed.

Comment (paplorinc) [reply subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractJavaGroovyCompileAvoidanceIntegrationSpec.groovy]: this is what I saw in the other test as well, but I agree it's better without: removed
